config/v1/types_infrastructure.go

```diff
@@ -110,6 +115,53 @@ const (
 	IBMCloudProviderTypeVPC IBMCloudProviderType = "VPC"
 )
 
+// PlatformSpec holds the desired state specific to the underlying infrastructure provider
+// of the current cluster. Since these are used at spec-level for the underlying cluster, it
+// is supposed that only one of the spec structs is set.
+type PlatformSpec struct {
+	// type is the underlying infrastructure provider for the cluster. This
+	// value controls whether infrastructure automation such as service load
+	// balancers, dynamic volume provisioning, machine creation and deletion, and
+	// other integrations are enabled. If None, no infrastructure automation is
+	// enabled. Allowed values are "AWS", "Azure", "BareMetal", "GCP", "Libvirt",
+	// "OpenStack", "VSphere", "oVirt", and "None". Individual components may not support
+	// all platforms, and must handle unrecognized platforms as None if they do
+	// not support that platform.
+	Type PlatformType `json:"type"`
```

Comment:
get up with @damemi or @sttts about formalizing union types

Reply:
@damemi @sttts can you provide some detail/pointers to achieve this? Thanks!

Reply:
make it required as in https://github.com/openshift/api/blob/master/config/v1/types_apiserver.go#L17 and follow https://github.com/openshift/api/blob/master/config/v1/types_apiserver.go#L96 for the sum type.

Reply:
What are valid values? Anything non-empty? You find the specification language here: https://book.kubebuilder.io/reference/markers/crd-validation.html

Reply:
> if you look at the document, this field needs to support new platform strings to support devel of new platforms and therefore, the restriction to certain set is not possible..

I think it is.  The doc is for the consumers, not for the struct itself.  We can restrict to a known set of of platforms.  It's the consumers that need to accept anything.